R/od-funs.R

```diff
@@ -406,6 +410,29 @@ line2pointsn.Spatial <- function(l) {
 line2pointsn.sf <- function(l) {
   suppressWarnings(sf::st_cast(l, "POINT"))
 }
+
+#' @rdname line2points
+#' @export
+line2vertices <- function(l) {
+  UseMethod("line2vertices")
+}
+#' @export
+line2vertices.sf <- function(l) {
+  all_vertexes <- sf::st_coordinates(l)
+  indexes_of_internal_vertexes <- lapply(
+    split(1:nrow(all_vertexes), all_vertexes[, "L1"]),
+    function(x) x[-c(1, length(x))] # exclude starting and ending point
+  )
+  # extract those indexes
+  internal_vertexes <- all_vertexes[do.call("c", indexes_of_internal_vertexes), ]
+
+  # transform back to sf
+  internal_vertexes_sf <- sf::st_as_sf(data.frame(internal_vertexes),
+    coords = c("X", "Y"), crs = sf::st_crs(l)
+  )
+  internal_vertexes_sf
+}
```

Comment:
I think there are more efficient ways of doing this. But happy for now if it works.

Reply:
Heads-up @agila5 I've tested it and it seems to return points that are also nodes, like the red point at the bottom of the plot in the reprex below:

``` r
library(stplanr)
#> Registered S3 method overwritten by 'R.oo':
#>   method        from       
#>   throw.default R.methodsS3
l <- routes_fast_sf[2:4, ]
lpoints <- line2points(l)
lpoints2 <- line2pointsn(l)
lvertices <- line2vertices(l)
plot(l$geometry)
plot(sf::st_geometry(lpoints), pch = lpoints$id, cex = lpoints$id, col = "black", add = TRUE)
plot(lpoints2$geometry, add = TRUE)
plot(lvertices$geometry, col = "red", add = TRUE)
```

![](https://i.imgur.com/4TKcATv.png)

<sup>Created on 2019-10-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
I'll check!

Reply:
I already checked and your way is pretty fast 🔥 

I thought the `split()` was doing `st_split()` in a `lapply()` which was why I thought it would be slow. But tests show it's super fast.

Reply:
Alternative:

```r
line2vertices.sf2 <- function(l) {
  all_vertexes <- sf::st_coordinates(l)
  node_vertices <- sf::st_coordinates(line2points(l))
  all_vertexes <- as.data.frame(all_vertexes[, 1:2])
  node_vertices <- as.data.frame(node_vertices[, 1:2])
  all_vertices_in_nodes = is.na(match(data.frame(t(all_vertexes[, 1:2])), data.frame(t(node_vertices[, 1:2]))))
  internal_vertices <- all_vertexes[all_vertices_in_nodes, ]
  
  sf::st_as_sf(data.frame(internal_vertices), coords = c("X", "Y"), crs = sf::st_crs(l))
  
  # indexes_of_internal_vertexes <- lapply(
  #   split(1:nrow(all_vertexes), all_vertexes[, "L1"]),
  #   function(x) x[-c(1, length(x))] # exclude starting and ending point
  # )
  # # extract those indexes
  # internal_vertexes <- all_vertexes[do.call("c", indexes_of_internal_vertexes), ]
  # 
  # # transform back to sf
  # internal_vertexes_sf <- sf::st_as_sf(data.frame(internal_vertexes),
  #   coords = c("X", "Y"), crs = sf::st_crs(l)
  # )
  # internal_vertexes_sf
}
```

Gives

```
bench::mark(line2vertices(l), line2vertices.sf2(l), check = F)
# A tibble: 2 x 13
  expression              min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result
  <bch:expr>           <bch:> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>
1 line2vertices(l)     1.03ms 1.05ms      938.    56.3KB     6.34   444     3      473ms <df[,…
2 line2vertices.sf2(l)  3.4ms 3.47ms      281.    93.7KB     6.91   122     3      434ms <df[,…
# … with 3 more variables: memory <list>, time <list>, gc <list>
```